File: qqbot-module-api/src/main/java/com/mieai/qqbot/module/api/ModuleValidation.java
package com.mieai.qqbot.module.api;

final class ModuleValidation {
    private ModuleValidation() {}

    static String requireId(String value, String name) {
        String text = requireText(value, name, 128);
        if (!text.matches("[a-z][a-z0-9]*(?:[._-][a-z0-9]+)*")) {
            throw new IllegalArgumentException(name + " is invalid");
        }
        return text;
    }

    static String requireVersion(String value, String name) {
        String text = requireText(value, name, 64);
        if (!text.matches("(?:0|[1-9][0-9]*)\\.(?:0|[1-9][0-9]*)\\.(?:0|[1-9][0-9]*)(?:[-+][0-9A-Za-z.-]+)?")) {
            throw new IllegalArgumentException(name + " is not a semantic version");
        }
        return text;
    }

    static String requireText(String value, String name, int maximumLength) {
        if (value == null || value.isBlank()) {
            throw new IllegalArgumentException(name + " must not be blank");
        }
        String text = value.strip();
        if (text.length() > maximumLength) {
            throw new IllegalArgumentException(name + " is too long");
        }
        return text;
    }
}

File: qqbot-module-api/src/main/kotlin/com/mieai/qqbot/module/api/ModuleValidation.kt
package com.mieai.qqbot.module.api

internal object ModuleValidation {
    private val ID_PATTERN = Regex("[a-z][a-z0-9]*(?:[._-][a-z0-9]+)*")
    private val VERSION_PATTERN = Regex("(?:0|[1-9][0-9]*)\\.(?:0|[1-9][0-9]*)\\.(?:0|[1-9][0-9]*)(?:[-+][0-9A-Za-z.-]+)?")

    fun requireId(value: String?, name: String): String {
        val text = requireText(value, name, 128)
        require(ID_PATTERN.matches(text)) { "$name is invalid" }
        return text
    }

    fun requireVersion(value: String?, name: String): String {
        val text = requireText(value, name, 64)
        require(VERSION_PATTERN.matches(text)) { "$name is not a semantic version" }
        return text
    }

    fun requireText(value: String?, name: String, maximumLength: Int): String {
        require(!value.isNullOrBlank()) { "$name must not be blank" }
        val text = value.trim()
        require(value == text) { "$name must not have surrounding whitespace" }
        require(text.length <= maximumLength) { "$name is too long" }
        return text
    }
}
